### ISSUE TITLE
Calculate prices excluding tax, show them including tax

### ISSUE THREAD

**Issue by stefandoorn** [Sylius/Sylius#7833]
I'm currently having some custom calculators to calculate the price of variants based on the variant itself but also on some additional information supplied on the OrderItem (sizes). I calculated everything excluding taxes, and had set the 'is_included_in_price' setting for the tax category to 0. Works fine, until I figured out that on the front-end my prices are still shown without tax in the cart, while it should be including tax. The summary does show the tax, but it should be on the product line itself. Therefore I rewrote my calculators to include tax in it, but due to rounding errors it can happen the final price excluding tax is still one cent of. Can be fixed ofcourse, but it led me to do some searching on how it works.

I found this page (http://docs.sylius.org/en/latest/book/orders/taxation.html#taxrates-included-in-price) telling two scenarios: 

* Calculate price without tax, show it without tax and only show tax in summary
* Calculate price with tax, show it with tax and also show tax in summary

There is one missing scenario that makes sense to me:

* Calculate price without tax, show it with tax and show tax in summary

Most often it's way easier to do price calculations without tax to find the base price. Based on context, tax should be added on top of it. In the calculator itself, I prefer not to include dependencies on context aware things to make the calculations.

Am I overlooking a feature in the framework that already suits this method, or would this be something that could be added? I think it's a quite important one, especially when having to do complex pricing.

**Comment by psihius** [comment]
I definitely second the ability to show prices in the product listings with or without the tax depending on the some setting - I'm building a multi-channel store myself and I need to show prices with tax conditionally - people from some countries pay VAT tax, other's don't.
As taxes may change, it is inconvenient to set a price with tax included, especially when there are many channels.

The tax applying also should be available to be switched on a customer/session basis, as when customer switches the country he is from (via a selector, and we are going to do some GeoIP detection to make it more friendly), we need to be able to toggle to toggle if the price needs to be shown with or without tax . Right now it is not possible - you either enter prices with tax and tick "tax included in price" or you set prices without tax and person see's how much tax he is paying only in the cart/checkout.
For our region (Eastern Europe) displaying prices without a tax is a big "no-no" - our prices always reflect what customer must pay for the good with all taxes included. But when you sell to multiple countries, sometimes the customer does not pay the tax and we need to show them that price without the tax - ideally a person switches the country he is buying from and that handles the tax settings. I believe that current zone and tax functionality can handle that, but sylius, as far as I know, can only apply taxes channel based and there is no "customer's country" concept at the moment and taxes are re-calculated only when a shipping/billing country is selected.
Essentially, what I would like to see are 2 things:

# Ability to set a country for the customer and apply zoning based on that country before any shipping info is entered.
# Ability to toggle application of tax when displaying product prices. Maybe this will be solved via 1st by using zoning.

**Comment by mrjingles** [comment]
I am not sure if my problem fits in that scenario. I have set up "tax included" to not calculate my netto prices depending on tax rate. So everything is fine, when the customer can see tax amount in cart. But when the customer selects a tax free shipping method (I added a VAT field for that to trigger if that method is available) the cart shows still the full price with included tax. So imho the price should be reduced with calculated tax amount, right?

**Comment by stefandoorn** [comment]
I think this is a different scenario, as it's mainly about showing it until the cart/checkout without any knowledge about the customer's location. I want to calculate excluding tax, but show it including tax. Yours might be quite a special case - probably EU VAT stuff - but not sure what is the best way to solve your issue.

**Comment by mrjingles** [comment]
But also after selecting shipping address and shipping method (without tax) on last checkout page the prices are not excluded tax, but tax position is 0.00. That doesn't make sense to me, but maybe I should open separate issue for that.

**Comment by stefandoorn** [comment]
True. The thing is that my issue is more about the calculation of product prices on a lower level when using a custom product calculator. It's separate indeed, as this is more about whether to apply tax or not based on certain scenario. Both make sense, as this is probably something regarding US -> EU trade or the other way around e.g.

**Comment by mrjingles** [comment]
@stefandoorn I am going your way now. Calculating without tax, but want to show it in frontend with tax included. Do you already have a solution that works for you?

**Comment by stefandoorn** [comment]
@mrjingles I went with a Twig helper to show the including VAT price based on the product / variant settings in the correct templates. Had to rewrite many of them anyway for our custom needs, so it was not a that big deal.

**Comment by mrjingles** [comment]
@stefandoorn I tried with a custom ProductVariantPriceCalculator, but this didn't work very well. Tax is still added in cart, when I have non EU shipment.
So can you give me a link to that twig helper and how to add it in templates?

**Comment by stefandoorn** [comment]
@mrjingles This seems to be different than my use case. I wanted to calculate prices without tax, and show it including - anywhere. We only ship at 21% VAT, period. So for me it was rather easy, as I don't need to have any clue about the customers location. I'm not sure how you can get around with your use case, as it seems to depend on customer context.

**Comment by pjedrzejewski** [comment]
Just like with tax categories for customer, we decided to tackle this after 1.0. It should be doable without BC breaks by introducing new Twig helpers that calculate proper amounts.

**Comment by rowanvanekeren** [comment]
@pjedrzejewski  is there any progression with this? It's pretty relevant for my use case aswell

**Comment by vvasiloi** [comment]
@rowanvanekeren in the meantime you can create a Twig helper like that: https://gist.github.com/vvasiloi/a381b4313d672abf4492cdbd1b36fb9d

**Comment by vvasiloi** [comment]
Do not stale.

**Comment by michalhosna** [comment]
Do not stale.

**Comment by vvasiloi** [comment]
Just adding some Twig helpers might not be enough, since the only way to get the tax zone right now, outside the checkout, is the channel's default tax zone.
Letting the customer choose a country isn't consistent with the way tax is calculated during checkout, so it can yield different results.
To keep it consistent, the customer will have to choose from the defined [tax] zones.
It might be better to implement different use-cases as plugins, instead of a general implementation in Sylius core.

**Comment by CoderMaggie** [comment]
![image](https://user-images.githubusercontent.com/13198869/147089029-3083f13d-c5eb-41c3-ad30-87117f43918f.png)

Hello! Long time no update on this issue, while in Sylius we've had major changes regarding the order summaries and taxes display (last year, I believe).

How do you feel after these changes? Is anything more still required?

**Comment by lchrusciel** [comment]
Should be fixed together with: https://github.com/Sylius/Sylius/issues/10261

**Comment by vvasiloi** [comment]
@CoderMaggie  @lchrusciel this issue is not about displaying the prices in the admin section, but in the shop section, specifically outside the checkout process, i.e. the catalog.
This is especially important for B2B shops.
However, it's a complex problem... as it is always with taxes.
As I said before, it's probably better to isolate the uses cases and provide dedicated solutions through plugins, instead of trying to solve it in the Sylius core.

**Comment by Fakrae** [comment]
As some additional information for why/when this would be usefull/needed, check OSS-laws from Mid 2021: https://vat-one-stop-shop.ec.europa.eu/select-language?destination=/node/1
Basically most if not all European countries need prices for private customers to be displays including taxes, but taxes depend on the delivery-address of the customer (different taxrates for the same product in different countries) so victors twig-helper would be the bare minimum (along with a way for the customer to set the country of delivery before entering the checkout/on the first visit).
So I'd argue, that it is even more relevant for B2C shops than for B2B - in B2B you could get away with displaying the prices without VAT.

I think in some countries it would theoretically be sufficient to display the totals during checkout, but the customer would not expect that and will not like/understand that the carts total changes after he enters his address...

**Comment by stefandoorn** [comment]
@GSadee @jakubtobiasz I want to bring your attention to this one for v2.0. I think this is pretty crucial.

**Comment by stefandoorn** [comment]
Is this something that will be covered for `v2.0`? It can imagine that it's such a breaking change that if not done in `v2.0`, it has to be in `v3.0`. 

cc @GSadee @jakubtobiasz @lchrusciel @vvasiloi @CoderMaggie @Zales0123

**Comment by vvasiloi** [comment]
If no one is working on it and if the ideas I shared previously seem like a good starting point, then I could commit to implementing it in 2.0, or maybe even 1.14. I see this mentioned quite often, so there seems to be enough interest in it.

**Comment by wadjeroudi** [comment]
@vvasiloi can you be more precise about your implementation plz ? previously, you advocated for an external solution via plugins.

**Comment by vvasiloi** [comment]
> @vvasiloi can you be more precise about your implementation plz ? previously, you advocated for an external solution via plugins.

I shared a gist in one of the comments. I still think it's too difficult to provide in the core a solution that would be generic enough and also cover many use-cases, at least for now. However, we could start by providing some basic tools to calculate and display prices with and without taxes. Maybe later that could  evolve into a solid foundation and maybe even some specific implementatios for the most common use-cases.

**Comment by Fakrae** [comment]
I recently payed BitBag a few thousand euros for a solution (switchable delivery-country in homepage-header, switch in channel to turn it on/off, message in checkout and cart about mismatching countries when logged in), I will try to extract the necessary changes but I'm not "fluent in git" and they are all over the place

**Comment by vvasiloi** [comment]
> I recently payed BitBag a few thousand euros for a solution (switchable delivery-country in homepage-header, switch in channel to turn it on/off, message in checkout and cart about mismatching countries when logged in), I will try to extract the necessary changes but I'm not "fluent in git" and they are all over the place

I implemented a country selector on [https://coolgift.com](coolgift.com) ~5 years ago, so I have the know-how, but it would be great to see how others did it.
